Allow transitions and animations to work within iframes

jacwright opened the pull request:
Svelte works great in iframes except for transitions and animations. This fixes that issue.

Fixes #3624.

Rich-Harris commented on src/runtime/internal/style_manager.ts:
```diff
@@ -1,9 +1,11 @@
+type DocStyles = [CSSStyleSheet, Record<string, true>];
+
+let activeDocs = new Set<Document>();
+let docStyles = new WeakMap<Document, DocStyles>();
```
We're not using `WeakMap` anywhere else currently, and I think we probably want that to continue being the case since it's basically impossible to polyfill in environments that don't support it. Reckon we could get away with this being a `Map` instead? In the vast majority of cases we're only going to have a single key, and it's always likely to be stable, so *maybe* it doesn't matter if we hang on to references?

Rich-Harris commented on src/runtime/internal/style_manager.ts:
```diff
@@ -25,14 +27,18 @@ export function create_rule(node: Element & ElementCSSInlineStyle, a: number, b:
+		doc.head.appendChild(style);
+		stylesheet = style.sheet as CSSStyleSheet;
+		docStyles.set(doc, [ stylesheet, current_rules ]);
+	}
```
This is obnoxious code-golfing, but if you wanted to save 18 bytes (after minification) you could do this...

```typescript
const doc = node.ownerDocument;
activeDocs.add(doc);
const [stylesheet, current_rules] = (docStyles.has(doc) ? docStyles : docStyles.set(doc, [
	(doc.head.appendChild(element('style') as HTMLStyleElement).sheet as CSSStyleSheet),
	{}
])).get(doc) as DocStyles;
```

Can we change `docStyles` to `doc_styles`?

Rich-Harris commented on src/runtime/internal/style_manager.ts:
```diff
@@ -53,14 +59,19 @@ export function delete_rule(node: Element & ElementCSSInlineStyle, name?: string
 		.join(', ');
 
-	if (name && !--active) clear_rules();
+	active = Math.max(0, active - 1);
```
This feels like it's masking a bug — is the behaviour different when iframes are involved, or is this unrelated to iframes?

jacwright replied:
Yes, it _is_ masking a bug. Toggling a transition on and off multiple times has a memory leak.

<img width="1321" alt="Screen Shot 2019-11-19 at 4 37 43 PM" src="https://user-images.githubusercontent.com/120294/69196159-ea6f2900-0aea-11ea-85f7-ad0c72bb78c6.png">

Rich-Harris commented on src/runtime/internal/style_manager.ts:
```diff
@@ -53,14 +59,19 @@ export function delete_rule(node: Element & ElementCSSInlineStyle, name?: string
+			while (i--) stylesheet.deleteRule(i);
+			docStyles.set(doc, [ stylesheet, {} ]);
+		});
+		activeDocs.clear();
```
oh! we can *definitely* use a `Map` instead of a `WeakMap` if we're clearing `activeDocs` (can we change it to `active_docs`?)

jtormey commented on src/runtime/internal/style_manager.ts:
```diff
@@ -53,14 +55,19 @@ export function delete_rule(node: Element & ElementCSSInlineStyle, name?: string
 
-	if (name && !--active) clear_rules();
+	active = Math.max(0, active - 1);
+	if (name && !active) clear_rules();
```
I've been working off of this branch (since my Svelte app needs iframe support) and noticed that this change causes a previously working transition of mine to get cut short mid-transition.

It's more noticeable with a longer transition duration. For example, with duration=5s, the first 2s or so of the transition work as expected, but then it jumps to the final position suddenly. This is the case both in and outside of an iframe.

My use case is pretty simple, just: `<div transition:slide={...}>`. Reverting just the above 3 lines fixes the issue for me, although then the leak is still present of course.

jacwright commented:
I've updated the PR to latest Svelte, fixed memory leaks by solidifying the logic in `delete_rules` while still allowing it to be called multiple times and getting rid of the Map which really needed to be a WeakMap. Storing the necessary data on the document object is a poor man's WeakMap.

@jtormey I would love it you could test this in your situation. If it cuts transitions short, it is because of a bug in `internal/transitions.ts` where it may be calling `delete_rules` too aggressively. I did not feel comfortable enough with that code to understand if/where it might be (i.e. I did not want to spend the time fixing any possible bugs in there for this PR). The logic in `style_manager` is correct now and deletes the rules when no elements have their `__svelte*` animations on them anymore.

Ready for another review @Rich-Harris. :D

jtormey commented:
Just retested with your latest changes, the transition no longer gets cut short! 🎉

jacwright commented:
Thanks @Conduitry! 🎉
